### PR TITLE
Update AutoML to use objective decision function during scoring for custom objectives

### LINKED ISSUE #1868: Automl does not use objective decision function during scoring
Our classification pipeline uses the `_compute_predictions` method in `score` to get the predicted labels and probabilities. 

However, `_compute_predictions` [does not pass the objective](https://github.com/alteryx/evalml/blob/main/evalml/pipelines/classification_pipeline.py#L138) to the pipeline `predict` method.  

Is this intended?

### PR DESCRIPTION
**Pull request by angela97lin**
Closes #1868

- Added `BinaryClassificationPipelineMixin` to share logic between binary classification / time series binary classification pipelines regarding the pipeline threshold.
- Moved `optimize_threshold` from ClassificationPipeline to BinaryClassificationPipelineMixin. I believe this makes sense since only binary classification pipelines, not multiclass classification pipelines, know about thresholds.

### CODE REVIEW

File: evalml/pipelines/binary_classification_pipeline.py
Comment (angela97lin) [inline evalml/pipelines/binary_classification_pipeline.py:48]: Helper function that takes in `predict_proba`, so we don't have to recalculate it for each objective.

File: evalml/pipelines/binary_classification_pipeline.py
Comment (chukarsten) [inline evalml/pipelines/binary_classification_pipeline.py:98]: Hmmm, the naming of the variables  and the try/except kind of implied to me that you were going to be forgiving of one or any of the objectives failing.  But I guess what you're trying to do is just get a catalog of what worked and what didn't without raising at the *first* instance of a failed objective.  Am I following that?  So either you get the scores only if everything scores correctly or you get and error detailing what failed and what was successful.

File: evalml/pipelines/binary_classification_pipeline.py
Comment (chukarsten) [inline evalml/pipelines/binary_classification_pipeline.py:80]: Is this pad something that we always want on or that the user would always expect to be True?

**Review by chukarsten — CHANGES_REQUESTED** [review]
I think there's something funky with the double anys on L77 of binary_classification_pipeline.py.  Besides that, this looks pretty solid.

File: evalml/pipelines/binary_classification_pipeline.py
Comment (angela97lin) [reply evalml/pipelines/binary_classification_pipeline.py:77]: Heh, yeah this is confusing stuff, good stuff to clarify. Hopefully this helps (and points out any errors I might have missed). Here's the parred-down version of the code in main for `_predict()`:

```
    def _predict(self, X, objective=None):
        if self.threshold is None:
            return self._component_graph.predict(X)
        ypred_proba = self.predict_proba(X).to_dataframe()
        ypred_proba = ypred_proba.iloc[:, 1]
        if objective is None:
            return infer_feature_types(ypred_proba > self.threshold)
        return infer_feature_types(objective.decision_function(ypred_proba, threshold=self.threshold, X=X))

```

Basically, if we have a `threshold` set for our pipeline, we use `predict_proba` and then either call `ypred_proba > self.threshold` or the decision function if passed an objective.

This logic here for computing the predictions means is trying to say:
1. Calculate predict_proba if any objective needs it for scoring (old behavior)
2. If objective can optimize threshold (aka score doesn't need probability--see `can_optimize_threshold` property) and the threshold has been set, then we will still want the predict probability to calculate the last two lines.

To your confusion @chukarsten:
1. If we only have an objective w/ score_needs_proba: we calculate (conditional is True)
2. If we only have an objective that has an optimizable threshold and threshold has been set: we calculate
3. If we only have an objective that has an optimizable threshold and threshold has not been set: we don't need to calculate (will call `self._component_graph.predict(X)` line above)

Writing this up though, maybe I can simplify this to `if any(o.score_needs_proba for o in objectives) or self.threshold is not None)` since `any(not o.score_needs_proba for o in objectives)` is captured in setting the threshold 🤔

File: evalml/pipelines/binary_classification_pipeline.py
Comment (angela97lin) [reply evalml/pipelines/binary_classification_pipeline.py:98]: Correct! This is actually mostly copied from `PipelineBase` and our current behavior, with lines 91-93 being the difference. But yes, you're spot on. We don't want to error out on the first objective scoring, and want to calculate what we can. We decided this because for AutoML, maybe some objectives just happen to error out but you don't really care about them. In that case, it doesn't make sense to halt search just for this objective, so best to just call that error out and then move on. (When *all* objectives fail though in AutoML though, we scream)

File: evalml/pipelines/binary_classification_pipeline.py
Comment (angela97lin) [reply evalml/pipelines/binary_classification_pipeline.py:80]: I think for time-series, yes! @freddyaboulton correct me if I'm wrong 😅

File: evalml/pipelines/binary_classification_pipeline.py
Comment (angela97lin) [reply evalml/pipelines/binary_classification_pipeline.py:77]: @chukarsten 
Okay, trying to wrap my head around this some more, and I guess the benefit of having `if any(o.score_needs_proba for o in objectives) or self.threshold is not None)` is that you only calculate if the threshold is set. In the case where score_needs_proba is False and the threshold is None, you don't need to calculate.


```
score needs proba | threshold is None | calculate predict_proba?
------------------------------------------------------------------------
           T               T                   YES
------------------------------------------------------------------------
           T               F                   YES
------------------------------------------------------------------------
           F               T                   YES
------------------------------------------------------------------------
           F               F                   NO

```

File: evalml/pipelines/binary_classification_pipeline.py
Comment (freddyaboulton) [reply evalml/pipelines/binary_classification_pipeline.py:80]: Yes! It's an internal detail of time series pipelines but we always want the predicted probabilities and predictions to be padded. @angela97lin Can you add coverage for time series pipelines to `test_binary_predict_pipeline_use_objective`?

File: evalml/pipelines/binary_classification_pipeline.py
Comment (freddyaboulton) [reply evalml/pipelines/binary_classification_pipeline.py:77]: @angela97lin Maybe I'm confused now too 😂  but shouldn't the last two 'YES' and 'NO' be flipped? If the threshold is None and the score does not need probability, you don't need to calculate predict_proba?

Overall, I'm in favor of rewritting in terms of `if any(o.score_needs_proba for o in objectives) or self.threshold is not None)`

Comment (angela97lin) [reply evalml/pipelines/binary_classification_pipeline.py:77]: @freddyaboulton I think I'm the confused one, haha--fixed! But sounds good, I'll simplify this 😁

File: evalml/pipelines/binary_classification_pipeline.py
Comment (freddyaboulton) [reply evalml/pipelines/binary_classification_pipeline.py:98]: I think it'd be best if there was only one `_score_all_objectives` implementation because I want the `PipelineScoreError` logic to only live in one place if we decide to change it in the future.

This works and I don't want to hold up the merge longer but I think we should at least file an issue for refactoring in the future. What do you think @angela97lin @chukarsten ?

**Review by freddyaboulton — APPROVED** [review]
@angela97lin This looks good to me! I left some scattered comments throughout. One thing that would be nice to address before merge is adding a time series test to `test_binary_predict_pipeline_use_objective`

File: evalml/pipelines/binary_classification_pipeline.py
Comment (angela97lin) [inline evalml/pipelines/binary_classification_pipeline.py:12]: Moved all of this to `BinaryClassificationPipelineMixin`

File: evalml/pipelines/binary_classification_pipeline.py
Comment (angela97lin) [reply evalml/pipelines/binary_classification_pipeline.py]: Did some more refactoring magic, so this is all in one place again 🎈

File: evalml/tests/pipeline_tests/test_pipelines.py
Comment (angela97lin) [inline evalml/tests/pipeline_tests/test_pipelines.py:2020]: Removed multiclass tests and moved to test_binary_classification_pipelines.py

**Pull request merged**
